### ISSUE TITLE
Saver producing large meta files after long run

### ISSUE THREAD

**Issue by domluna** [tensorflow/tensorflow#2654]
The code is for deep q-learning and for reference it's [here](https://github.com/domluna/deep-rl-gym-tutorials/tree/master/q_learning) and the keras model is [here](https://github.com/domluna/deep-rl-gym-tutorials/blob/master/q_learning/models.py)

Training after 400 timesteps:

save file is 9.7MB
meta file is 25MB

after 2,750,000 timesteps

save file is 9.7MB
meta file is 1.4GB

after 3M timesteps

save file is 9.7MB
meta file is 1.5GB

So the meta file grows over 100MB per 275,000 timesteps. The saver keeps the last 5 checkpoints by default so you could image my shock when I woke up and my Dropbox was full!

Anyway, this kind of growth is unreasonable is there a way to keep it in check?

**Comment by girving** [comment]
@domluna: Assuming the meta file is a MetaGraphDef, can you check whether it's the `GraphDef` part that's growing?  One possibility is that you are adding new ops to the graph each step, causing the graph to grow linearly.

**Comment by mrry** [comment]
FYI, if you aren't intending to use the `MetaGraphDef`, you can call `saver.save(..., write_meta_graph=False)` to avoid writing it with each checkpoint.

**Comment by domluna** [comment]
Assuming I've done this right, using https://gist.github.com/domluna/f06921196b539a397a6a24ae7f571276 to check the ByteSize. I get the following step/size outputs:

20000, 30376060
30000, 35425130
40000, 40474200
50000, 45523270
60000, 50572340

So `GraphDef` is growing, just not sure why ...

**Comment by girving** [comment]
@domluna: Try printing out the node names in the graph each step.  If you see it growing over time, it probably means your adding new nodes to the graph over time.

**Comment by yaroslavvb** [comment]
You can do "tf.get_default_graph().finalize()" after you've done modifying
your graph, that'll throw an error on any new modifications

**Comment by domluna** [comment]
So an update on this. The `finalize` method is great I'm using it on all my code now to make sure nothing sketchy is going on.

I think I found the problem with the particular code and it might be an issue with Keras. When I call `train_on_batch` on a Keras model.

The culprit being 

https://github.com/fchollet/keras/blob/a1610eb2747e2020b9460b1f60e6e750d253fffb/keras/engine/training.py#L673

After some more digging it appears `K.gradients` and in turn `tf.gradients` is being called each time this is called.

**Comment by girving** [comment]
@fchollet: Is there a Keras issue here?

**Comment by fchollet** [comment]
@domluna no, if you read the code you linked, you can see that `tf.gradients` is called exactly once, then the results are cached in the model instance (`train_function` attribute). Calling `train_on_batch` many times only results in a single call to `tf.gradients` at the very beginning.

Since `Model` cannot be modified after being instantiated, this cache is never invalidated.

@girving as far as I can tell, no.

**Comment by domluna** [comment]
@fchollet my mistake, thanks for the explanation.
